### ISSUE TITLE
const fn type is missing `const` in error message

### ISSUE THREAD

**Issue by varkor** [rust-lang/rust#57527]
```rust
const fn foo() {}

fn main() {
    let _: () = foo;
}
// error[E0308]: mismatched types
// = note: expected type `()`
//             found type `fn() {foo}`
```
I'd expect the found type to be `const fn() {foo}`.

**Comment by Centril** [comment]
```rust
const fn foo() {}

fn main() {
    let _: const fn() = foo;
}
```

There's no such type `const fn()`, so saying this would be misleading.

The current diagnostic seems correct to me.

**Comment by varkor** [comment]
`fn()` is a function pointer type. `fn() {foo}`, on the other hand, is a unique function type. It can't be directly by the user. Analogously, even though there is no `const fn()` function pointer type, it still makes sense to describe the unique function types corresponding to `const fn` with a preceding `const`.

**Comment by Centril** [comment]
> `fn() {foo}`, on the other hand, is a unique function type. It can't be directly by the user.

Yes, but I think the purpose of showing `fn()` here even tho the actual type is the unnameable ZST `{foo}` is that the user can take `fn()` and use it. If you show `const fn() {foo}` you take away this piece of information that is actionable, thereby making the diagnostic less useful.

**Comment by estebank** [comment]
Current output:

```
error[E0308]: mismatched types
 --> src/main.rs:4:17
  |
1 | const fn foo() {}
  | -------------- fn() {foo} defined here
...
4 |     let _: () = foo;
  |                 ^^^
  |                 |
  |                 expected (), found fn item
  |                 help: use parentheses to call this function: `foo()`
  |
  = note: expected type `()`
             found type `fn() {foo}`
```

I don't know what team this falls under. `T-lang`?, `T-doc`?

**Comment by Centril** [comment]
Diagnostics fall under t-compiler.
